ci/ci/build.py

```diff
@@ -92,40 +92,39 @@ def __init__(self, code, config_str, scope, *, requested_step_names=(), excluded
             raise BuildConfigurationError('Excluding build steps is only permitted in a dev scope')
 
         config = yaml.safe_load(config_str)
-        name_step: Dict[str, Optional[Step]] = {}
-        self.steps: List[Step] = []
-
         if requested_step_names:
             log.info(f"Constructing build configuration with steps: {requested_step_names}")
 
+        all_steps: List[Step] = []
+        name_step: Dict[str, Step] = {}
         for step_config in config['steps']:
-            step_params = StepParameters(code, scope, step_config, name_step)
-            step = Step.from_json(step_params)
-            if not step.run_if_requested or step.name in requested_step_names:
-                self.steps.append(step)
-                name_step[step.name] = step
-            else:
-                name_step[step.name] = None
-
-        # transitively close requested_step_names over dependencies
+            step = Step.from_json(StepParameters(code, scope, step_config, name_step))
+            name_step[step.name] = step
+            all_steps.append(step)
+
         if requested_step_names:
+            # transitively close requested_step_names over dependencies
             visited = set()
 
-            def request(step: Step):
+            def visit_dependent(step: Step):
                 if step not in visited and step.name not in excluded_step_names:
+                    if not step.can_run_in_current_cloud():
+                        raise BuildConfigurationError(f'Step {step.name} cannot be run in cloud {CLOUD}')
                     visited.add(step)
                     for s2 in step.deps:
-                        request(s2)
+                        visit_dependent(s2)
 
             for step_name in requested_step_names:
-                request(name_step[step_name])
-            self.steps = [s for s in self.steps if s in visited]
+                visit_dependent(name_step[step_name])
+            self.steps = [step for step in all_steps if step in visited]
+        else:
+            self.steps = [step for step in all_steps if not step.run_if_requested and step.can_run_in_current_cloud()]
 
     def build(self, batch, code, scope):
         assert scope in ('deploy', 'test', 'dev')
 
         for step in self.steps:
-            if (step.scopes is None or scope in step.scopes) and (step.clouds is None or CLOUD in step.clouds):
+            if step.can_run_in_scope(scope):
```

Comment:
let's add an assert that it can run in current cloud, the logic is currently nonlocal and thus a bit hard to follow.